### PR TITLE
[Sketcher] Angle constraint refactor

### PR DESCRIPTION
**Pull request by PaddleStroke**
First step to solve https://github.com/FreeCAD/FreeCAD/issues/10524

This PR reduce code duplicity in the command angle and in sketcher_dimension.

### CODE REVIEW

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -8500,6 +8500,83 @@ double SketchObject::calculateAngleViaPoint(int GeoId1, int GeoId2, double px, d
         throw Base::ValueError("Null geometry in calculateAngleViaPoint");
 }
 
+bool SketchObject::calculateAngle(int& GeoId1, int& GeoId2, Sketcher::PointPos& PosId1, Sketcher::PointPos& PosId2, double& ActAngle)
+{
+    const Part::Geometry* geom1 = getGeometry(GeoId1);
+    const Part::Geometry* geom2 = getGeometry(GeoId2);
+
+    if (geom1->getTypeId() == Part::GeomLineSegment::getClassTypeId() &&
+        geom2->getTypeId() == Part::GeomLineSegment::getClassTypeId()) {
```
[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:8509]: You're right!

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -8500,6 +8500,84 @@ double SketchObject::calculateAngleViaPoint(int GeoId1, int GeoId2, double px, d
...
```
Comment (benj5378) [inline src/Mod/Sketcher/App/SketchObject.cpp:8509]: You changed the behaviour when trying to make the guard clause. You didn't invert the condition

Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Ah this is unfortunate. Fixed

**Comment by howie-j** [comment]
Perhaps include arc angle also? https://github.com/FreeCAD/FreeCAD/issues/10598

**Comment by PaddleStroke** [comment]
@howie-j I'll make a separate PR as it's not really related to this PR.

**Comment by PaddleStroke** [comment]
@abdullahtahiriyo interested by this little refactor perhaps?
CI complains because of precommit.

**Comment by PaddleStroke** [comment]
Build fails but I don't think it's due to this PR?

**Comment by PaddleStroke** [comment]
@wwmayer care to take a quick look here ? It's just a little refactor because some code was copy-pasted in 3 places in CommandConstraint.

File: src/Mod/Sketcher/App/SketchObject.h [hunk 3]
```diff
@@ -472,6 +472,12 @@ class SketcherExport SketchObject: public Part::Part2DObject
     static void appendMalformedConstraintsMsg(const std::vector<int>& malformed, std::string& msg);
 
     double calculateAngleViaPoint(int geoId1, int geoId2, double px, double py);
+    // Find the angle between two lines. Return false if geoIds are not lines.
+    bool calculateAngle(int& geoId1,
```
Comment (wwmayer) [inline src/Mod/Sketcher/App/SketchObject.h:476]: the method should be declared as `const` if possible.

**Comment by wwmayer** [comment]
Please also consider the reported lint warnings (clang-diagnostic error, modernize C++, ...).

**Comment by PaddleStroke** [comment]
I applied const and added #include <Gui/Command.h> that was giving a clang error.
Thanks

**Comment by PaddleStroke** [comment]
Got it. Please let me know if the forward declaration is correctly done.

**Comment by wwmayer** [comment]
Yes, it's fine.

**Comment by abdullahtahiriyo** [comment]
@PaddleStroke 

Why does the member function added to SketchObject need to be in SketchObject? In other words, why it is not a free function in CommandConstraints.h instead?

I am tracking the dependencies visually, so I may be wrong, but it seems to me that that member function does not access anything private or protected of SketchObject. It is not used by any other part of the App namespace either, only Gui.  

SketchObject is already some kind of God object. To avoid it becoming unmaintainable, we should strive to only add there what is necessary. 

When functionality needed in App is to be added and there is no need for a member function, but a function using the interface is enough, one could consider either a free function in App namespace, or a new class (see SketchAnalysis as an example).

If for any reason, there is no dependency in SketchObject and there is no need for it in App namespace and CommandConstraints.h is not a good place (for example, because you will use it elsewhere in Gui), which I fail to see at this moment, then perhaps Sketcher/Gui/Utils.h could be an alternative place.

Let me know if in doubt.

**Comment by PaddleStroke** [comment]
There is no necessity. I did that because I was following `calculateAngleViaPoint` pattern. I thought making code sharable was the correct way of doing things. Especially when what this function is doing (calculating angle between two lines) could potentially be used in other places of sketcher.

**Comment by abdullahtahiriyo** [comment]
calculateAngleViaPoint could be refactored out of SketchObject. Probably the reason it has not been done yet is that it is used by the tangency locking function in SketchObject. It may be possible to refactor that one out too (but within App), as it is needed for the constraint addition mechanism.

It is definitely good to reuse code. As a free function, it is equally reusable (you just need to add the header where you want to reuse it). 

As for scope, it is not extremely strict. There is some leeway. I would go for the broadest where it makes sense now and avoid broader where YAGNI. For example, here it must be within the Sketcher module, because it makes use of Sketcher module classes (meaning not global App/Gui). Nowadays, I think it is only used in SketcherGui. Eventually it might perhaps be possible to use it at Sketcher namespace level. However, looking at what Sketcher namespace does now, it looks to me like a YAGNI scenario. For this reason, I would tend to maintain it within SketcherGui. If this is just used as a refactor of constraints, I think it makes sense to put it in CommandConstraints.

I am interested in exposing the rationale so that we understand why we do things one way and not the other.

**Comment by PaddleStroke** [comment]
It's clear. So I will move it in CommandConstraint.

**Comment by abdullahtahiriyo** [comment]
Let me know when you are happy with the code.

**Comment by PaddleStroke** [comment]
I moved the function. Please confirm if this is ok.

**Pull request merged**